File: lifeos/static/js/calendar-api.js
/**
 * LifeOS Calendar - API Module
 * Handles calendar events and interpretations CRUD operations.
 * 
 * Usage:
 *   const { listEvents, createEvent, confirmInterpretation } = window.lifeosCalendar;
 *   const events = await listEvents({ start_date: '2025-12-01', end_date: '2025-12-31' });
 *   const event = await createEvent({ title: 'Gym workout', start_time: '2025-12-07T18:00:00' });
 * 
 * @module calendar-api
 */
(function () {
  "use strict";

  const API_BASE = "/api/calendar";

  /**
   * Build auth headers from lifeosAuth with CSRF token from meta tag
   * The meta tag CSRF is session-bound and more reliable than localStorage
   * @param {Object} [extra={}] - Additional headers to merge
   * @returns {HeadersInit}
   */
  function authHeaders(extra = {}) {
    let base = {};
    if (window.lifeosAuth && typeof window.lifeosAuth.authHeaders === "function") {
      base = window.lifeosAuth.authHeaders(extra);
    } else {
      base = { "Content-Type": "application/json", ...extra };
    }
    // Prefer CSRF token from meta tag (matches current session) over localStorage
    const metaCsrf = document.querySelector('meta[name="csrf-token"]')?.content;
    if (metaCsrf) base["X-CSRF-Token"] = metaCsrf;
    return base;
  }

  // ==================== Calendar Events ====================

  /**
   * List calendar events with optional filters
   * @param {Object} [options] - Query options
   * @param {string} [options.start_date] - ISO datetime start filter
   * @param {string} [options.end_date] - ISO datetime end filter
   * @param {string} [options.source] - Source filter (manual, sync_google, sync_apple, api)
   * @param {number} [options.limit=50] - Max results
   * @param {number} [options.offset=0] - Pagination offset
   * @returns {Promise<Array>} List of calendar events
   */
  async function listEvents(options = {}) {
    const params = new URLSearchParams();
    if (options.start_date) params.set("start_date", options.start_date);
    if (options.end_date) params.set("end_date", options.end_date);
    if (options.source) params.set("source", options.source);
    if (options.limit) params.set("limit", String(options.limit));
    if (options.offset) params.set("offset", String(options.offset));

    const url = `${API_BASE}/events${params.toString() ? "?" + params : ""}`;
    const resp = await fetch(url, {
      method: "GET",
      headers: authHeaders(),
      credentials: "include",
    });

    const json = await resp.json();
    if (!json.ok) {
      throw new Error(json.error || "Failed to fetch events");
    }
    return json.events;
  }

  /**
   * Get a single calendar event by ID
   * @param {number} eventId - Event ID
   * @returns {Promise<Object>} Event with interpretations
   */
  async function getEvent(eventId) {
    const resp = await fetch(`${API_BASE}/events/${eventId}`, {
      method: "GET",
      headers: authHeaders(),
      credentials: "include",
    });

    const json = await resp.json();
    if (!json.ok) {
      throw new Error(json.error || "Failed to fetch event");
    }
    return json.event;
  }

  /**
   * Create a new calendar event
   * @param {Object} data - Event data
   * @param {string} data.title - Event title (required)
   * @param {string} data.start_time - ISO datetime (required)
   * @param {string} [data.end_time] - ISO datetime
   * @param {string} [data.description] - Event description
   * @param {string} [data.location] - Location
   * @param {boolean} [data.all_day=false] - All-day event
   * @param {string} [data.color] - Hex color
   * @param {boolean} [data.is_private=false] - Privacy flag
   * @param {Array<string>} [data.tags=[]] - Tags
   * @returns {Promise<Object>} Created event
   */
  async function createEvent(data) {
    const resp = await fetch(`${API_BASE}/events`, {
      method: "POST",
      headers: authHeaders({ "Content-Type": "application/json" }),
      credentials: "include",
      body: JSON.stringify(data),
    });

    const json = await resp.json();
    if (!json.ok) {
      throw new Error(json.error || "Failed to create event");
    }
    return json.event;
  }

  /**
   * Update an existing calendar event
   * @param {number} eventId - Event ID
   * @param {Object} data - Fields to update
   * @returns {Promise<Object>} Updated event
   */
  async function updateEvent(eventId, data) {
    const resp = await fetch(`${API_BASE}/events/${eventId}`, {
      method: "PATCH",
      headers: authHeaders({ "Content-Type": "application/json" }),
      credentials: "include",
      body: JSON.stringify(data),
    });

    const json = await resp.json();
    if (!json.ok) {
      throw new Error(json.error || "Failed to update event");
    }
    return json.event;
  }

  /**
   * Delete a calendar event
   * @param {number} eventId - Event ID
   * @returns {Promise<void>}
   */
  async function deleteEvent(eventId) {
    const resp = await fetch(`${API_BASE}/events/${eventId}`, {
      method: "DELETE",
      headers: authHeaders(),
      credentials: "include",
    });

    const json = await resp.json();
    if (!json.ok) {
      throw new Error(json.error || "Failed to delete event");
    }
  }

  // ==================== Interpretations ====================

  /**
   * Get pending interpretations for review
   * @param {Object} [options] - Query options
   * @param {string} [options.domain] - Filter by domain
   * @param {number} [options.limit=50] - Max results
   * @param {number} [options.offset=0] - Pagination offset
   * @returns {Promise<Array>} List of pending interpretations
   */
  async function getPendingInterpretations(options = {}) {
    const params = new URLSearchParams();
    if (options.domain) params.set("domain", options.domain);
    if (options.limit) params.set("limit", String(options.limit));
    if (options.offset) params.set("offset", String(options.offset));

    const url = `${API_BASE}/interpretations/pending${params.toString() ? "?" + params : ""}`;
    const resp = await fetch(url, {
      method: "GET",
      headers: authHeaders(),
      credentials: "include",
    });

    const json = await resp.json();
    if (!json.ok) {
      throw new Error(json.error || "Failed to fetch interpretations");
    }
    return json.interpretations;
  }

  /**
   * Update interpretation status
   * @param {number} interpretationId - Interpretation ID
   * @param {string} status - New status: 'confirmed', 'rejected', 'ignored'
   * @returns {Promise<Object>} Updated interpretation
   */
  async function updateInterpretation(interpretationId, status) {
    const resp = await fetch(`${API_BASE}/interpretations/${interpretationId}`, {
      method: "PATCH",
      headers: authHeaders({ "Content-Type": "application/json" }),
      credentials: "include",
      body: JSON.stringify({ status }),
    });

    const json = await resp.json();
    if (!json.ok) {
      throw new Error(json.error || "Failed to update interpretation");
    }
    return json.interpretation;
  }

  /**
   * Confirm an interpretation (creates domain record)
   * @param {number} interpretationId - Interpretation ID
   * @returns {Promise<Object>} Updated interpretation
   */
  async function confirmInterpretation(interpretationId) {
    return updateInterpretation(interpretationId, "confirmed");
  }

  /**
   * Reject an interpretation
   * @param {number} interpretationId - Interpretation ID
   * @returns {Promise<Object>} Updated interpretation
   */
  async function rejectInterpretation(interpretationId) {
    return updateInterpretation(interpretationId, "rejected");
  }

  /**
   * Ignore an interpretation
   * @param {number} interpretationId - Interpretation ID
   * @returns {Promise<Object>} Updated interpretation
   */
  async function ignoreInterpretation(interpretationId) {
    return updateInterpretation(interpretationId, "ignored");
  }

  // ==================== Utilities ====================

  /**
   * Get events for a specific date
   * @param {Date|string} date - Date to fetch events for
   * @returns {Promise<Array>} Events for that day
   */
  async function getEventsForDate(date) {
    const d = typeof date === "string" ? new Date(date) : date;
    const dateStr = d.toISOString().split("T")[0];
    return listEvents({
      start_date: dateStr + "T00:00:00",
      end_date: dateStr + "T23:59:59",
    });
  }

  /**
   * Get events for current week
   * @returns {Promise<Array>} Events for this week
   */
  async function getEventsThisWeek() {
    const today = new Date();
    const weekStart = new Date(today);
    weekStart.setDate(today.getDate() - today.getDay());
    const weekEnd = new Date(weekStart);
    weekEnd.setDate(weekStart.getDate() + 6);

    return listEvents({
      start_date: weekStart.toISOString(),
      end_date: weekEnd.toISOString(),
    });
  }

  /**
   * Get pending interpretation count
   * @param {string} [domain] - Optional domain filter
   * @returns {Promise<number>} Count of pending interpretations
   */
  async function getPendingCount(domain = null) {
    const interpretations = await getPendingInterpretations({ domain, limit: 500 });
    return interpretations.length;
  }

  // ==================== Domain Constants ====================

  const DOMAINS = {
    finance: { name: "Finance", icon: "💰", color: "#0b4f9c" },
    health: { name: "Health", icon: "❤️", color: "#dc2626" },
    habits: { name: "Habits", icon: "🎯", color: "#7c3aed" },
    skills: { name: "Skills", icon: "📚", color: "#059669" },
    projects: { name: "Projects", icon: "📁", color: "#ea580c" },
    relationships: { name: "Relationships", icon: "👥", color: "#db2777" },
  };

  const SOURCES = {
    manual: "Manual",
    sync_google: "Google Calendar",
    sync_apple: "Apple Calendar",
    api: "API",
  };

  // Public API
  const lifeosCalendar = {
    // Events
    listEvents,
    getEvent,
    createEvent,
    updateEvent,
    deleteEvent,
    // Interpretations
    getPendingInterpretations,
    updateInterpretation,
    confirmInterpretation,
    rejectInterpretation,
    ignoreInterpretation,
    // Utilities
    getEventsForDate,
    getEventsThisWeek,
    getPendingCount,
    // Constants
    DOMAINS,
    SOURCES,
  };

  // Expose to window
  window.lifeosCalendar = lifeosCalendar;

  // ES module export (if supported)
  if (typeof module !== "undefined" && module.exports) {
    module.exports = lifeosCalendar;
  }
})();

File: lifeos/domains/calendar/controllers/calendar_api.py
"""Calendar API controllers."""

from __future__ import annotations

from flask import Blueprint, jsonify, redirect, request
from flask_jwt_extended import get_jwt_identity, jwt_required

from lifeos.core.utils.decorators import csrf_protected, require_roles
from lifeos.domains.calendar.schemas import (
    CalendarEventCreate,
    CalendarEventListParams,
    CalendarEventUpdate,
    InterpretationUpdate,
)
from lifeos.domains.calendar.services import (
    create_calendar_event,
    delete_calendar_event,
    get_calendar_event,
    get_pending_interpretations,
    list_calendar_events,
    update_calendar_event,
    update_interpretation_status,
)
from lifeos.domains.calendar.services.google_sync_service import (
    GoogleCalendarError,
    disconnect_google_calendar,
    exchange_code_for_tokens,
    get_authorization_url,
    get_sync_status,
    save_oauth_token,
    sync_google_calendar,
)
from lifeos.extensions import limiter

calendar_api_bp = Blueprint("calendar_api", __name__)


# ==================== Calendar Events ====================


@calendar_api_bp.get("/events")
@jwt_required()
@limiter.limit("240/minute")
def list_events():
    """
    List calendar events with optional date range filter.
    
    Query Parameters:
    - start_date: ISO datetime (optional)
    - end_date: ISO datetime (optional)
    - source: 'manual', 'sync_google', 'sync_apple', 'api' (optional)
    - limit: max results (default 50, max 500)
    - offset: pagination offset (default 0)
    """
    user_id = int(get_jwt_identity())

    try:
        params = CalendarEventListParams.model_validate(request.args)
    except Exception:
        return jsonify({"ok": False, "error": "validation_error"}), 400

    events = list_calendar_events(
        user_id=user_id,
        start_date=params.start_date,
        end_date=params.end_date,
        source=params.source,
        limit=params.limit,
        offset=params.offset,
    )

    return jsonify({
        "ok": True,
        "events": [
            {
                "id": e.id,
                "title": e.title,
                "description": e.description,
                "start_time": e.start_time.isoformat(),
                "end_time": e.end_time.isoformat() if e.end_time else None,
                "all_day": e.all_day,
                "location": e.location,
                "source": e.source,
                "color": e.color,
                "is_private": e.is_private,
                "tags": e.tags,
                "duration_minutes": e.duration_minutes,
                "created_at": e.created_at.isoformat(),
            }
            for e in events
        ],
    }), 200


@calendar_api_bp.get("/events/<int:event_id>")
@jwt_required()
@limiter.limit("240/minute")
def get_event(event_id: int):
    """Get a single calendar event with its interpretations."""
    user_id = int(get_jwt_identity())

    event = get_calendar_event(user_id, event_id)
    if not event:
        return jsonify({"ok": False, "error": "not_found"}), 404

    return jsonify({
        "ok": True,
        "event": {
            "id": event.id,
            "title": event.title,
            "description": event.description,
            "start_time": event.start_time.isoformat(),
            "end_time": event.end_time.isoformat() if event.end_time else None,
            "all_day": event.all_day,
            "location": event.location,
            "source": event.source,
            "external_id": event.external_id,
            "color": event.color,
            "is_private": event.is_private,
            "tags": event.tags,
            "duration_minutes": event.duration_minutes,
            "created_at": event.created_at.isoformat(),
            "updated_at": event.updated_at.isoformat(),
            "interpretations": [
                {
                    "id": i.id,
                    "domain": i.domain,
                    "record_type": i.record_type,
                    "record_id": i.record_id,
                    "confidence_score": float(i.confidence_score),
                    "status": i.status,
                    "classification_data": i.classification_data,
                }
                for i in event.interpretations
            ],
        },
    }), 200


@calendar_api_bp.post("/events")
@jwt_required()
@csrf_protected
@require_roles({"calendar:write"})
@limiter.limit("120/minute")
def create_event():
    """
    Create a new calendar event.
    
    Request Body:
    {
      "title": "Lunch with John",
      "start_time": "2025-12-07T12:00:00",
      "end_time": "2025-12-07T13:00:00",
      "description": "Discuss project plans",
      "location": "Cafe XYZ",
      "all_day": false,
      "tags": ["meeting", "work"]
    }
    """
    user_id = int(get_jwt_identity())
    payload = request.get_json(silent=True) or {}

    try:
        data = CalendarEventCreate.model_validate(payload)
    except Exception:
        return jsonify({"ok": False, "error": "validation_error"}), 400

    try:
        event = create_calendar_event(
            user_id=user_id,
            title=data.title,
            start_time=data.start_time,
            end_time=data.end_time,
            description=data.description,
            location=data.location,
            all_day=data.all_day,
            color=data.color,
            is_private=data.is_private,
            tags=data.tags,
            metadata=data.metadata,
        )

        return jsonify({
            "ok": True,
            "event": {
                "id": event.id,
                "title": event.title,
                "start_time": event.start_time.isoformat(),
                "end_time": event.end_time.isoformat() if event.end_time else None,
                "created_at": event.created_at.isoformat(),
            },
        }), 201

    except ValueError as exc:
        code = str(exc)
        if code in {"invalid_title", "title_too_long"}:
            return jsonify({"ok": False, "error": code}), 400
        return jsonify({"ok": False, "error": "validation_error"}), 400


@calendar_api_bp.patch("/events/<int:event_id>")
@jwt_required()
@csrf_protected
@require_roles({"calendar:write"})
@limiter.limit("120/minute")
def update_event(event_id: int):
    """Update an existing calendar event."""
    user_id = int(get_jwt_identity())
    payload = request.get_json(silent=True) or {}

    try:
        data = CalendarEventUpdate.model_validate(payload)
    except Exception:
        return jsonify({"ok": False, "error": "validation_error"}), 400

    try:
        event = update_calendar_event(
            user_id=user_id,
            event_id=event_id,
            title=data.title,
            description=data.description,
            start_time=data.start_time,
            end_time=data.end_time,
            all_day=data.all_day,
            location=data.location,
            color=data.color,
            is_private=data.is_private,
            tags=data.tags,
            metadata=data.metadata,
        )

        return jsonify({
            "ok": True,
            "event": {
                "id": event.id,
                "title": event.title,
                "updated_at": event.updated_at.isoformat(),
            },
        }), 200

    except ValueError as exc:
        code = str(exc)
        if code == "not_found":
            return jsonify({"ok": False, "error": "not_found"}), 404
        if code == "invalid_title":
            return jsonify({"ok": False, "error": code}), 400
        return jsonify({"ok": False, "error": "validation_error"}), 400


@calendar_api_bp.delete("/events/<int:event_id>")
@jwt_required()
@csrf_protected
@require_roles({"calendar:write"})
@limiter.limit("60/minute")
def delete_event(event_id: int):
    """Delete a calendar event."""
    user_id = int(get_jwt_identity())

    try:
        delete_calendar_event(user_id, event_id)
        return jsonify({"ok": True}), 200
    except ValueError as exc:
        code = str(exc)
        if code == "not_found":
            return jsonify({"ok": False, "error": "not_found"}), 404
        return jsonify({"ok": False, "error": "validation_error"}), 400


# ==================== Interpretations ====================


@calendar_api_bp.get("/interpretations/pending")
@jwt_required()
@limiter.limit("240/minute")
def list_pending_interpretations():
    """
    List pending interpretations awaiting user review.
    
    Query Parameters:
    - domain: filter by domain (optional)
    - limit: max results (default 50)
    - offset: pagination offset (default 0)
    """
    user_id = int(get_jwt_identity())
    domain = request.args.get("domain")
    limit = int(request.args.get("limit", 50))
    offset = int(request.args.get("offset", 0))

    interpretations = get_pending_interpretations(
        user_id=user_id,
        domain=domain,
        limit=min(limit, 500),
        offset=offset,
    )

    return jsonify({
        "ok": True,
        "interpretations": [
            {
                "id": i.id,
                "calendar_event_id": i.calendar_event_id,
                "domain": i.domain,
                "record_type": i.record_type,
                "confidence_score": float(i.confidence_score),
                "status": i.status,
                "classification_data": i.classification_data,
                "created_at": i.created_at.isoformat(),
            }
            for i in interpretations
        ],
    }), 200


@calendar_api_bp.patch("/interpretations/<int:interpretation_id>")
@jwt_required()
@csrf_protected
@require_roles({"calendar:write"})
@limiter.limit("120/minute")
def update_interpretation(interpretation_id: int):
    """
    Update interpretation status (confirm, reject, ignore).
    
    Request Body:
    {
      "status": "confirmed" | "rejected" | "ignored"
    }
    """
    user_id = int(get_jwt_identity())
    payload = request.get_json(silent=True) or {}

    try:
        data = InterpretationUpdate.model_validate(payload)
    except Exception:
        return jsonify({"ok": False, "error": "validation_error"}), 400

    try:
        interpretation = update_interpretation_status(
            user_id=user_id,
            interpretation_id=interpretation_id,
            status=data.status,
        )

        return jsonify({
            "ok": True,
            "interpretation": {
                "id": interpretation.id,
                "status": interpretation.status,
                "updated_at": interpretation.updated_at.isoformat(),
            },
        }), 200

    except ValueError as exc:
        code = str(exc)
        if code == "not_found":
            return jsonify({"ok": False, "error": "not_found"}), 404
        if code == "invalid_status":
            return jsonify({"ok": False, "error": code}), 400
        return jsonify({"ok": False, "error": "validation_error"}), 400


# ==================== Google Calendar OAuth ====================


@calendar_api_bp.get("/oauth/google/authorize")
@jwt_required()
def google_oauth_authorize():
    """
    Initiate Google Calendar OAuth flow.
    
    Returns a redirect URL for the user to authorize access.
    """
    user_id = int(get_jwt_identity())
    auth_url = get_authorization_url(user_id, state=str(user_id))
    return jsonify({"ok": True, "authorization_url": auth_url}), 200


@calendar_api_bp.get("/oauth/google/callback")
def google_oauth_callback():
    """
    Handle Google OAuth callback.
    
    Exchanges code for tokens and redirects to calendar page.
    """
    code = request.args.get("code")
    state = request.args.get("state")
    error = request.args.get("error")
    
    if error:
        return redirect(f"/calendar?error={error}")
    
    if not code or not state:
        return redirect("/calendar?error=invalid_callback")
    
    try:
        user_id = int(state)
        token_data = exchange_code_for_tokens(code)
        save_oauth_token(user_id, token_data)
        return redirect("/calendar?connected=google")
    except (ValueError, GoogleCalendarError) as e:
        return redirect(f"/calendar?error={str(e)[:50]}")


@calendar_api_bp.get("/oauth/google/status")
@jwt_required()
def google_oauth_status():
    """
    Get Google Calendar connection status.
    """
    user_id = int(get_jwt_identity())
    status = get_sync_status(user_id)
    return jsonify({"ok": True, **status}), 200


@calendar_api_bp.post("/oauth/google/disconnect")
@jwt_required()
@csrf_protected
@require_roles({"calendar:write"})
def google_oauth_disconnect():
    """
    Disconnect Google Calendar integration.
    """
    user_id = int(get_jwt_identity())
    disconnected = disconnect_google_calendar(user_id)
    return jsonify({"ok": True, "disconnected": disconnected}), 200


@calendar_api_bp.post("/sync/google")
@jwt_required()
@csrf_protected
@require_roles({"calendar:write"})
@limiter.limit("10/minute")
def trigger_google_sync():
    """
    Manually trigger Google Calendar sync.
    """
    user_id = int(get_jwt_identity())
    
    try:
        stats = sync_google_calendar(user_id)
        return jsonify({"ok": True, **stats}), 200
    except GoogleCalendarError as e:
        return jsonify({"ok": False, "error": str(e)}), 400


# ==================== Apple Calendar (CalDAV) ====================


@calendar_api_bp.post("/oauth/apple/connect")
@jwt_required()
@csrf_protected
@require_roles({"calendar:write"})
def apple_connect():
    """
    Connect Apple Calendar using app-specific password.
    
    Request Body:
    {
        "apple_id": "user@icloud.com",
        "app_password": "xxxx-xxxx-xxxx-xxxx"
    }
    
    Users must generate an app-specific password at:
    https://appleid.apple.com/account/manage
    """
    from lifeos.domains.calendar.services.apple_sync_service import (
        AppleCalendarError,
        save_apple_credentials,
        verify_apple_credentials,
    )
    
    user_id = int(get_jwt_identity())
    payload = request.get_json() or {}
    
    apple_id = payload.get("apple_id", "").strip()
    app_password = payload.get("app_password", "").strip()
    
    if not apple_id or not app_password:
        return jsonify({"ok": False, "error": "apple_id and app_password required"}), 400
    
    # Verify credentials
    if not verify_apple_credentials(apple_id, app_password):
        return jsonify({"ok": False, "error": "invalid_credentials"}), 401
    
    # Save credentials
    save_apple_credentials(user_id, apple_id, app_password)
    
    return jsonify({"ok": True, "message": "Apple Calendar connected"}), 200


@calendar_api_bp.get("/oauth/apple/status")
@jwt_required()
def apple_status():
    """
    Get Apple Calendar connection status.
    """
    from lifeos.domains.calendar.services.apple_sync_service import get_apple_sync_status
    
    user_id = int(get_jwt_identity())
    status = get_apple_sync_status(user_id)
    return jsonify({"ok": True, **status}), 200


@calendar_api_bp.post("/oauth/apple/disconnect")
@jwt_required()
@csrf_protected
@require_roles({"calendar:write"})
def apple_disconnect():
    """
    Disconnect Apple Calendar integration.
    """
    from lifeos.domains.calendar.services.apple_sync_service import disconnect_apple_calendar
    
    user_id = int(get_jwt_identity())
    disconnected = disconnect_apple_calendar(user_id)
    return jsonify({"ok": True, "disconnected": disconnected}), 200


@calendar_api_bp.post("/sync/apple")
@jwt_required()
@csrf_protected
@require_roles({"calendar:write"})
@limiter.limit("10/minute")
def trigger_apple_sync():
    """
    Manually trigger Apple Calendar sync.
    """
    from lifeos.domains.calendar.services.apple_sync_service import (
        AppleCalendarError,
        sync_apple_calendar,
    )
    
    user_id = int(get_jwt_identity())
    
    try:
        stats = sync_apple_calendar(user_id)
        return jsonify({"ok": True, **stats}), 200
    except AppleCalendarError as e:
        return jsonify({"ok": False, "error": str(e)}), 400
